### PR TITLE
connection: prevent writing pending data for health checkers by introducing ConnectionCloseType::Abort

### PR DESCRIPTION
**Pull request by JuniorHsu**
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message:
Please check #23718

Additional Description:
Risk Level: low
Testing: unit test in connection and health checker
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #23718
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/network/connection_impl.cc [hunk 1]
```diff
@@ -140,7 +140,7 @@ void ConnectionImpl::close(ConnectionCloseType type) {
   const bool delayed_close_timeout_set = delayed_close_timeout_.count() > 0;
   if (data_to_write == 0 || type == ConnectionCloseType::NoFlush ||
       !transport_socket_->canFlushClose()) {
-    if (data_to_write > 0) {
+    if (data_to_write > 0 && type != ConnectionCloseType::NoFlush) {
       // We aren't going to wait to flush, but try to write as much as we can if there is pending
       // data.
       transport_socket_->doWrite(*write_buffer_, true);
```
[the remark this answers is not in the document]
Comment (JuniorHsu) [reply source/common/network/connection_impl.cc:146]: thanks for a quick update! i was trying to see if we have any current unit breakage. 
i'm not super sure what you mean about the else block.
we want to prevent `doWrite`  for NoFlush even if we have data, right?

File: test/common/network/connection_impl_test.cc [hunk 2]
```diff
@@ -2880,6 +2880,19 @@ TEST_F(PostCloseConnectionImplTest, NoReadAfterCloseFlushWriteWriteData) {
...
 }
 
+// Test that close(ConnectionCloseType::Abort) won't write and flush pending data.
+TEST_F(PostCloseConnectionImplTest, CloseAbort) {
+  InSequence s;
+  initialize();
+  writeSomeData();
+
+  // Connection abort. We have data written above in writeSomeData(),
+  // it won't be written and flushed due to Abort.
```
Comment (soulxu) [inline test/common/network/connection_impl_test.cc:2890]: super nit, s/Abort/abort

**Review by soulxu — DISMISSED** [review]
LGTM, thanks

File: test/common/network/connection_impl_test.cc (hunk 2 above)
Comment (JuniorHsu) [reply test/common/network/connection_impl_test.cc]: hmm `abort` might fail grammar  in some sense. I use ``ConnectionCloseType::Abort`` if you don't mind. Thanks for catching this!

File: changelogs/current.yaml [hunk 3]
```diff
@@ -107,6 +107,9 @@ bug_fixes:
 - area: jwt_authn
   change: |
     fix a bug that jwt_cache breaks the :ref:`provider_and_audiences <envoy_v3_api_field_extensions.filters.http.jwt_authn.v3.JwtRequirement.provider_and_audiences>` JWT requirement.
+- area: health_checker
+  change: |
+    prevent writing pending data for health checkers by introducing ``ConnectionCloseType::Abort``.
```
Comment (snowp) [inline changelogs/current.yaml:112]: For my understanding, why does this matter? What would happen previously? Should we include some detail in the release note?

**Review by snowp — CHANGES_REQUESTED** [review]
Overall lgtm, just one comment re release note. Is this something we'd want to runtime guard?

File: changelogs/current.yaml (hunk 3 above)
Comment (JuniorHsu) [reply changelogs/current.yaml:112]: The detail is in #23718. 
tl;dr there's cascading cpu backpressue for health checker handshake while cpu is throttled by docker cgroup limit. More and more HC requests are timeout-ed and closed before handshake, but we still try to TLS handshake.

IMO runtime guard is **not** needed for now since it doesn't matter to give up health checker's data.

However it needs some deliberation to replace as much `NoFlush` with `Abort` as possible in the future
https://github.com/envoyproxy/envoy/issues/23718#issuecomment-1341338352

I'll mention the detail in the release note in next commit.

File: envoy/network/connection.h [hunk 4]
```diff
@@ -69,8 +69,9 @@ class ConnectionCallbacks {
 enum class ConnectionCloseType {
   FlushWrite, // Flush pending write data before raising ConnectionEvent::LocalClose
   NoFlush,    // Do not flush any pending data and immediately raise ConnectionEvent::LocalClose
```
Comment (soulxu) [inline envoy/network/connection.h:71]: It would be better to mention that this type will give a try to write left data.

Comment (JuniorHsu) [reply envoy/network/connection.h]: addressed. thanks!

**Review by soulxu — COMMENTED** [review]
thanks, LGTM

**Review by yanavlasov — APPROVED** [review]

**Comment by JuniorHsu** [comment]
coverage ci failure should be unrelated given I didn't change the folder
```
Code coverage for source/extensions/http/early_header_mutation is lower than limit of 96.6 (95.2)
Code coverage for source/extensions/filters/network/http_connection_manager is lower than limit of 96.6 (96.4)
```

**Comment by JuniorHsu** [comment]
merge to top which mitigate coverage issue

**Comment by JuniorHsu** [comment]
ci passed and we got approved. could we merge this? Thanks. @snowp @yanavlasov

**Comment by yanavlasov** [comment]
I'll ping @snowp for LGTM and then submit.
